Sources/Workspace/DefaultPluginScriptRunner.swift

```
@@ -179,35 +182,80 @@ public struct DefaultPluginScriptRunner: PluginScriptRunner {
-                dispatchPrecondition(condition: .onQueue(callbackQueue))
-                completion($0.tryMap {
-                    // Emit the compiler output as observable info.
-                    let compilerOutput = ((try? $0.utf8Output()) ?? "") + ((try? $0.utf8stderrOutput()) ?? "")
-                    observabilityScope.emit(info: compilerOutput)
-
-                    // We return a PluginCompilationResult for both the successful and unsuccessful cases (to convey diagnostics, etc).
-                    let result = PluginCompilationResult(compilerResult: $0, diagnosticsFile: diagnosticsFile, compiledExecutable: executableFile)
-                    guard $0.exitStatus == .terminated(code: 0) else {
-                        throw DefaultPluginScriptRunnerError.compilationFailed(result)
-                    }
-                    return result
-                })
+            // Hash the command line and the contents of the source files to decide whether we need to recompile the plugin executable. We have a bounded set of input conditions, and can control what we hash.
+            let compilerInputsHash: String? = try? {
+                let stream = BufferedOutputByteStream()
+                stream <<< command
+                for sourceFile in sources.paths {
+                    try stream <<< localFileSystem.readFileContents(sourceFile).contents
+                }
+                return stream.bytes.sha256Checksum
+            }()
+
+            // If we already have a compiled executable, then compare its hash with the new one. If they're the same, we don't need to compile.
+            var compilationNeeded = true
+            let hashFile = executableFile.parentDirectory.appending(component: execName + ".inputhash")
+            if localFileSystem.exists(executableFile) {
+                if (try? localFileSystem.readFileContents(hashFile)) == compilerInputsHash {
```

is try? hiding error we should report?

---

Agreed.  Fixed.